gradio/queue.py

```diff
@@ -81,10 +98,24 @@ async def start_processing(self) -> None:
 
             # Using mutex to avoid editing a list in use
             async with self.delete_lock:
-                event = self.event_queue.pop(0)
-
-            self.active_jobs[self.active_jobs.index(None)] = event
-            run_coro_in_background(self.process_event, event)
+                first_event = self.event_queue.pop(0)
+                events = [first_event]
+
+                event_fn_index = first_event.fn_index
+                batch = self.blocks_dependencies[event_fn_index]["batch"]
+                batch_size = self.blocks_dependencies[event_fn_index]["max_batch_size"]
+
+                if batch:
+                    while len(events) < batch_size:
+                        fn_indices = [ev.fn_index for ev in self.event_queue]
+                        try:
+                            index = fn_indices.index(event_fn_index)
+                            events.append(self.event_queue.pop(index))
+                        except ValueError:
+                            break
```

Comment:
@FarukOzderim didn't you suggest not having separate queues for separate events because it would add unnecessary complexity in #2216? :D

I agree some of the "fill batch" logic could be eliminated by having separate queues. However, I think it would be replaced with similarly complex logic with the workers having to pick events from different queues, which is why we closed #2216. If I had to pick between the two, I think this approach is simpler

In addition, I'll rewrite this code to make the "fill batch" logic faster (especially since it's a blocking operation)

Reply:
@abidlabs did I do that? haha :D

The part that I was against was scheduling and complexity coming with it. One of the alternatives I suggested over there aligns with this current suggestion:

> 
Allocating special thread slots to events, a thread that will just serve to a single event. (OS already does scheduling between threads and this will move the complexity to the OS instead of Gradio. Honestly, It feels unnecessary to implement a scheduling in Gradio when OS already schedules the threads.

My view point in both suggestions is complexity, I think handling batched event in a different queue concurrently (without getting into any scheduling complexity) is more simpler and easier to implement&maintain. It is better to make use of easiness of Python coroutines. Though at the same time, it is not a very big difference. Let me know how these look to you @abidlabs @freddyaboulton.

Reply:
Actually, I think you are right, a single queue would be more simpler. Seeing searching for the next events part carried me to the suggestions above. Smt like this would be quite simple. There would be a wait list for each batched event.

```
batch_wait_list = []
...


while loop
...
event = queue.pop()

if event.batch:
  batch_wait_list.append(event)
  if batch_wait_duration > time_threshold:
    await process_batch(batch_wait_list);
  if len(batch_wait_list) < batch_size:
    continue
  else:
    await process_batch(batch_wait_list);

else:
  await process_event(event)
```

Reply:
Rewrote the code to be more efficient. I want to avoid using `time_threshold` for v1 of the queue.